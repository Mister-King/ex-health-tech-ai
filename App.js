/**
 * Health Tech AI Exercise
 *
 * @format
 * @flow strict-local
 */
import 'react-native-gesture-handler';
import React, { useState, useEffect } from 'react';
import axios from 'axios';

import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

import ScreenError from './components/ScreenError/ScreenError';
import ScreenLoading from './components/ScreenLoading/ScreenLoading';
import ScreenHome from './components/ScreenHome/ScreenHome';
import ScreenPost from './components/ScreenPost/ScreenPost';

import { PostsContext } from './components/Posts/PostsContext';

import styles from './_app.scss';

const Stack = createStackNavigator();

const App: () => React$Node = () => {
    // State
    const [isLoading, setIsLoading] = useState(true);
    const [data, setData] = useState(null);
    const [error, setError] = useState(null);
    const [activePostId, setActivePostId] = useState(null);

    // On mount
    useEffect(() => { fetchData(); }, []);

    // Internal functions
    const fetchData = async () => {
        let posts;
        let users;
        let comments;

        setIsLoading(true);
        setError(false);

        try {
            posts = await axios.get('http://jsonplaceholder.typicode.com/posts')
                .then(response => response.data);

            if (posts) {
                users = await axios.get('http://jsonplaceholder.typicode.com/users')
                    .then(response => response.data);
            }

            if (users) {
                comments = await axios.get('http://jsonplaceholder.typicode.com/comments')
                    .then(response => response.data);
            }

            setData({ posts, users, comments });
        } catch (error) {
            setError(true);
        } finally {
            setTimeout(() => setIsLoading(false), 300);
        }
    }

    const updateActivePost = id => {
        setActivePostId(id);
    }

    // Render
    // Error screen
    if (!isLoading && error) {
        return <ScreenError fetchData={fetchData} />;
    }

    // Main navigation screens
    if (!isLoading && data) {
        return (
            <PostsContext.Provider
                value={
                    {
                        updateActivePost,
                        activePostId,
                        data,
                    }
                }
            >
                <NavigationContainer>
                    <Stack.Navigator screenOptions={{
                        headerStyle: styles['nav-header']
                    }}>
                        <Stack.Screen
                            name="Posts"
                            component={ScreenHome}
                        />
                        <Stack.Screen
                            name="Post"
                            component={ScreenPost}
                        />
                    </Stack.Navigator>
                </NavigationContainer>
            </PostsContext.Provider>
        );
    }

    // Loading screen
    return <ScreenLoading />;
};

export default App;
