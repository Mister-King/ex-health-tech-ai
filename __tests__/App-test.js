/* eslint no-undef: 0 */
/**
 * @format
 */

import 'react-native';
import React from 'react';
import { create, act } from 'react-test-renderer';
import axios from 'axios';
import { PostsData, UsersData, CommentsData } from '../jest/jest.data';
import App from '../App';

jest.mock('axios');

test('renders activity spinner on load', () => {
    jest.useFakeTimers();

    const tree = create(<App/>).toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders error screen on API error', () => {
    jest.useFakeTimers();

    axios.get.mockImplementation(() => Promise.reject(new Error('not found')));

    let tree;

    act(() => {
        tree = create(<App/>);
        jest.runAllTimers();
    });

    expect(tree.toJSON()).toMatchSnapshot();
});

test('fetches data from API and renders posts', async () => {
    jest.useFakeTimers();

    axios.get.mockImplementation((url) => {
        switch (url) {
        case 'http://jsonplaceholder.typicode.com/posts':
            return Promise.resolve({ data: PostsData });
        case 'http://jsonplaceholder.typicode.com/users':
            return Promise.resolve({ data: UsersData });
        case 'http://jsonplaceholder.typicode.com/comments':
            return Promise.resolve({ data: CommentsData });
        default:
            return Promise.reject(new Error('not found'));
        }
    });

    let tree;

    act(() => {
        tree = create(<App />);
        jest.runAllTimers();
    });
    expect(tree.toJSON()).toMatchSnapshot();
});
