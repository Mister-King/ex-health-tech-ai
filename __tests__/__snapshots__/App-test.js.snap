// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches data from API and renders posts 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(216, 216, 216)",
                "flex": 1,
                "shadowColor": "rgb(216, 216, 216)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                aria-level="1"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                Posts
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          "screen",
                          "screen--trim",
                        ]
                      }
                    >
                      <View>
                        <RCTScrollView
                          ItemSeparatorComponent={[Function]}
                          data={
                            Array [
                              Object {
                                "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
                                "id": 1,
                                "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
                                "userId": 1,
                              },
                              Object {
                                "body": "est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
                                "id": 2,
                                "title": "qui est esse",
                                "userId": 1,
                              },
                            ]
                          }
                          disableVirtualization={false}
                          getItem={[Function]}
                          getItemCount={[Function]}
                          horizontal={false}
                          initialNumToRender={10}
                          keyExtractor={[Function]}
                          maxToRenderPerBatch={10}
                          onContentSizeChange={[Function]}
                          onEndReachedThreshold={2}
                          onLayout={[Function]}
                          onMomentumScrollEnd={[Function]}
                          onScroll={[Function]}
                          onScrollBeginDrag={[Function]}
                          onScrollEndDrag={[Function]}
                          removeClippedSubviews={false}
                          renderItem={[Function]}
                          scrollEventThrottle={50}
                          stickyHeaderIndices={Array []}
                          style="postlist"
                          updateCellsBatchingPeriod={50}
                          viewabilityConfigCallbackPairs={Array []}
                          windowSize={21}
                        >
                          <View>
                            <View
                              onLayout={[Function]}
                              style={null}
                            >
                              <View
                                accessible={true}
                                focusable={true}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onResponderGrant={[Function]}
                                onResponderMove={[Function]}
                                onResponderRelease={[Function]}
                                onResponderTerminate={[Function]}
                                onResponderTerminationRequest={[Function]}
                                onStartShouldSetResponder={[Function]}
                                style="postitem"
                              >
                                <Text
                                  style="postitem__text"
                                >
                                  sunt aut facere repellat provident occaecati excepturi optio reprehenderit
                                </Text>
                              </View>
                              <View
                                style="divider"
                              />
                            </View>
                            <View
                              onLayout={[Function]}
                              style={null}
                            >
                              <View
                                accessible={true}
                                focusable={true}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onResponderGrant={[Function]}
                                onResponderMove={[Function]}
                                onResponderRelease={[Function]}
                                onResponderTerminate={[Function]}
                                onResponderTerminationRequest={[Function]}
                                onStartShouldSetResponder={[Function]}
                                style="postitem"
                              >
                                <Text
                                  style="postitem__text"
                                >
                                  qui est esse
                                </Text>
                              </View>
                            </View>
                          </View>
                        </RCTScrollView>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;

exports[`renders activity spinner on load 1`] = `
<View
  style={
    Array [
      "screen",
      "screen--full",
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#19191f"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`renders error screen on API error 1`] = `
<View
  style={
    Array [
      "screen",
      "screen--full",
    ]
  }
>
  <Text
    style={
      Array [
        "headline",
      ]
    }
  >
    Oh no!
  </Text>
  <Text>
    Something went wrong. Try again in a bit.
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Try again
      </Text>
    </View>
  </View>
</View>
`;
